ant-networking/src/relay_manager.rs

```diff
@@ -239,3 +378,227 @@ impl RelayManager {
+                Some(Protocol::P2p(id)) => id,
+                _ => {
+                    debug!("Incoming connection does not have a valid 'from peer id'. Not tracking its health.");
+                    return;
+                }
+            }
+        };
+
+        match self.incoming_connections_from_remote_peer.entry(from_peer) {
+            Entry::Occupied(mut entry) => {
+                entry
+                    .get_mut()
+                    .push((relay_server, *connection_id, SystemTime::now(), None));
+            }
+            Entry::Vacant(entry) => {
+                entry.insert(vec![(
+                    relay_server,
+                    *connection_id,
+                    SystemTime::now(),
+                    None,
+                )]);
+            }
+        }
+    }
+
+    fn on_connection_established(&mut self, from_peer: &PeerId, connection_id: &ConnectionId) {
+        if let Some(connections) = self
+            .incoming_connections_from_remote_peer
+            .get_mut(from_peer)
+        {
+            if let Some((_, _, _, succeeded)) = connections
+                .iter_mut()
+                .find(|(_, id, _, _)| id == connection_id)
+            {
+                *succeeded = Some(true);
+            }
+        }
+
+        self.try_update_stat();
+    }
+
+    fn on_incomming_connection_error(
+        &mut self,
+        send_back_addr: &Multiaddr,
+        connection_id: &ConnectionId,
+    ) {
+        let from_peer = {
+            match send_back_addr
+                .iter()
+                .find(|p| matches!(p, Protocol::P2p(_)))
+            {
+                Some(Protocol::P2p(id)) => id,
+                _ => {
+                    debug!("Incoming connection does not have a valid 'from peer id'. Not tracking its health.");
+                    return;
+                }
+            }
+        };
+
+        if let Some(connections) = self
+            .incoming_connections_from_remote_peer
+            .get_mut(&from_peer)
+        {
+            if let Some((_, _, _, succeeded)) = connections
+                .iter_mut()
+                .find(|(_, id, _, _)| id == connection_id)
+            {
+                *succeeded = Some(false);
+            }
+        }
+
+        self.try_update_stat();
+    }
+
+    fn try_update_stat(&mut self) {
+        let mut to_remove = Vec::new();
+
+        for (from_peer, connections) in self.incoming_connections_from_remote_peer.iter_mut() {
+            let Some(latest_time) = connections.iter().map(|(_, _, time, _)| time).max() else {
+                debug!("The incoming connections from {from_peer:?} are empty. Skipping.");
+                continue;
+            };
+
+            let Ok(elapsed) = SystemTime::now().duration_since(*latest_time) else {
+                debug!("Could not obtain elapsed time.");
+                continue;
+            };
+
+            if elapsed < MAX_DURATION_TO_TRACK_INCOMING_CONNECTIONS_PER_PEER {
```

Comment:
not quite sure, MAX_DURATION_TO_TRACK_INCOMING_CONNECTIONS_PER_PEER is a fixed duration.
However, according to the log msg, `waiting to be established` is an event driven status change, not `time-driven`
Don't get why `not timed out` means it is still waiting to be established?